File: app/src/main/java/de/grobox/transportr/trips/search/TripViewHolder.java
/*    Transportr
 *    Copyright (C) 2013 - 2017 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.grobox.transportr.trips.search;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import de.grobox.transportr.R;
import de.grobox.transportr.trips.search.TripAdapter.OnTripClickListener;
import de.schildbach.pte.dto.Trip;
import de.schildbach.pte.dto.Trip.Individual;
import de.schildbach.pte.dto.Trip.Leg;
import de.schildbach.pte.dto.Trip.Public;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static de.grobox.transportr.utils.DateUtils.getDelayText;
import static de.grobox.transportr.utils.DateUtils.getDuration;
import static de.grobox.transportr.utils.DateUtils.getTime;
import static de.grobox.transportr.utils.TransportrUtils.addLineBox;
import static de.grobox.transportr.utils.TransportrUtils.addWalkingBox;
import static de.grobox.transportr.utils.TransportrUtils.getLocationName;
import static de.grobox.transportr.utils.TransportrUtils.setRelativeDepartureTime;

class TripViewHolder extends RecyclerView.ViewHolder {

	private final View root;
	private final TextView fromTimeRel;
	private final TextView fromTime, fromLocation, fromDelay;
	private final TextView toTime, toLocation, toDelay;
	private final ViewGroup lines;
	private final TextView duration;

	TripViewHolder(View v) {
		super(v);
		root = v;
		fromTimeRel = v.findViewById(R.id.fromTimeRel);
		fromTime = v.findViewById(R.id.fromTime);
		fromLocation = v.findViewById(R.id.fromLocation);
		fromDelay = v.findViewById(R.id.fromDelay);
		toTime = v.findViewById(R.id.toTime);
		toLocation = v.findViewById(R.id.toLocation);
		toDelay = v.findViewById(R.id.toDelay);
		lines = v.findViewById(R.id.lines);
		duration = v.findViewById(R.id.duration);
	}

	void bind(final Trip trip, OnTripClickListener listener) {
		// Relative Departure Time
		setRelativeDepartureTime(fromTimeRel, trip.getFirstDepartureTime());

		// Departure Time
		Leg firstLeg = trip.legs.get(0);
		if (firstLeg instanceof Public) {
			fromTime.setText(getTime(fromTime.getContext(), ((Public) firstLeg).getDepartureTime(true)));
		} else {
			fromTime.setText(getTime(fromTime.getContext(), firstLeg.getDepartureTime()));
		}

		// Departure Delay
		Public publicLeg = trip.getFirstPublicLeg();
		if (publicLeg != null && publicLeg.getDepartureDelay() != null && publicLeg.getDepartureDelay() != 0) {
			fromDelay.setText(getDelayText(publicLeg.getDepartureDelay()));
			fromDelay.setVisibility(VISIBLE);
		} else {
			fromDelay.setVisibility(GONE);
		}
		fromLocation.setText(getLocationName(trip.from));

		// Lines and Trip Duration
		lines.removeAllViews();
		for (Leg leg : trip.legs) {
			if (leg instanceof Public) {
				addLineBox(lines.getContext(), lines, ((Public) leg).line);
			} else if (leg instanceof Individual) {
				addWalkingBox(lines.getContext(), lines);
			} else {
				throw new RuntimeException();
			}
		}
		duration.setText(getDuration(trip.getDuration()));

		// Arrival Time
		Leg lastLeg = trip.legs.get(trip.legs.size() - 1);
		if (lastLeg instanceof Public) {
			toTime.setText(getTime(toTime.getContext(), ((Public) lastLeg).getDepartureTime(true)));
		} else {
			toTime.setText(getTime(toTime.getContext(), lastLeg.getDepartureTime()));
		}

		// Arrival Delay
		publicLeg = trip.getLastPublicLeg();
		if (publicLeg != null && publicLeg.getArrivalDelay() != null && publicLeg.getArrivalDelay() != 0) {
			toDelay.setText(getDelayText(publicLeg.getArrivalDelay()));
			toDelay.setVisibility(VISIBLE);
		} else {
			toDelay.setVisibility(GONE);
		}
		toLocation.setText(getLocationName(trip.to));

		// Click Listener
		root.setOnClickListener(view -> listener.onClick(trip));
	}

}

File: app/src/main/java/de/grobox/transportr/trips/search/TripViewHolder.kt
/*
 *    Transportr
 *
 *    Copyright (c) 2013 - 2021 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.grobox.transportr.trips.search

import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.TextView
import com.google.common.base.Strings.isNullOrEmpty
import de.grobox.transportr.R
import de.grobox.transportr.trips.BaseViewHolder
import de.grobox.transportr.trips.detail.TripUtils.getStandardFare
import de.grobox.transportr.trips.detail.TripUtils.hasFare
import de.grobox.transportr.trips.search.TripAdapter.OnTripClickListener
import de.grobox.transportr.ui.LineView
import de.grobox.transportr.utils.DateUtils.formatDuration
import de.grobox.transportr.utils.DateUtils.formatTime
import de.grobox.transportr.utils.DateUtils.formatRelativeTime
import de.grobox.transportr.utils.TransportrUtils.getLocationName
import de.schildbach.pte.dto.Trip
import de.schildbach.pte.dto.Trip.Individual
import de.schildbach.pte.dto.Trip.Public

internal class TripViewHolder(private val v: View) : BaseViewHolder(v) {

    private val fromTimeRel: TextView = v.findViewById(R.id.fromTimeRel)
    private val fromLocation: TextView = v.findViewById(R.id.fromLocation)
    private val toLocation: TextView = v.findViewById(R.id.toLocation)
    private val warning: View = v.findViewById(R.id.warning)
    private val lines: ViewGroup = v.findViewById(R.id.lines)
    private val duration: TextView = v.findViewById(R.id.duration)
    private val price: TextView = v.findViewById(R.id.price)

    fun bind(trip: Trip, listener: OnTripClickListener) {
        if (trip.isTravelable) {
            formatRelativeTime(fromTimeRel.context, trip.firstDepartureTime).let {
                fromTimeRel.apply {
                    text = it.relativeTime
                    visibility = it.visibility
                }
            }
        } else {
            fromTimeRel.setText(R.string.trip_not_travelable)
            fromTimeRel.visibility = VISIBLE
        }

        // Departure Time and Delay
        val firstLeg = trip.legs[0]
        if (firstLeg is Public) {
            setDepartureTimes(fromTime, fromDelay, firstLeg.departureStop)
        } else {
            fromTime.text = formatTime(context, firstLeg.departureTime)
            val firstPublicLeg = trip.firstPublicLeg
            if (firstPublicLeg != null && firstPublicLeg.departureDelay != null && firstPublicLeg.departureDelay != 0L) {
                setDepartureTimes(null, toDelay, firstPublicLeg.departureStop)
            }
        }
        fromLocation.text = getLocationName(trip.from)

        // Lines
        lines.removeAllViews()
        for (leg in trip.legs) {
            val lineView = LayoutInflater.from(context).inflate(R.layout.list_item_line, lines, false) as LineView
            when (leg) {
                is Public -> lineView.setLine(leg.line)
                is Individual -> lineView.setWalk()
                else -> throw RuntimeException()
            }
            lines.addView(lineView)
        }

        // Warning and Duration
        warning.visibility = if (trip.hasProblem()) VISIBLE else GONE
        duration.text = formatDuration(trip.duration)
        price.visibility = if (trip.hasFare()) VISIBLE else GONE
        price.text = trip.getStandardFare()

        // Arrival Time and Delay
        val lastLeg = trip.legs[trip.legs.size - 1]
        if (lastLeg is Public) {
            setArrivalTimes(toTime, toDelay, lastLeg.arrivalStop)
        } else {
            toTime.text = formatTime(context, lastLeg.arrivalTime)
            val lastPublicLeg = trip.lastPublicLeg
            if (lastPublicLeg != null && lastPublicLeg.arrivalDelay != null && lastPublicLeg.arrivalDelay != 0L) {
                setArrivalTimes(null, toDelay, lastPublicLeg.arrivalStop)
            }
        }
        toLocation.text = getLocationName(trip.to)

        // Click Listener
        v.setOnClickListener { listener.onClick(trip) }
    }

    private fun Trip.hasProblem(): Boolean {
        if (!isTravelable) return true
        for (leg in legs) {
            if (leg !is Public) continue
            if (!isNullOrEmpty(leg.message)) return true
            if (!isNullOrEmpty(leg.line?.message)) return true
        }
        return false
    }

}
